### ISSUE TITLE
Multidimensional arrays don't work

### ISSUE THREAD

**Issue by wengooooo** [CuyZ/Valinor#453]
```
final class SomeClass
{
    // If I don't comment then it works fine
    //public readonly string $description;

    /** @var Identifier[] */
    public array $identifier;

}

final class Identifier
{
    public string $value;
}

$mapper = (new \CuyZ\Valinor\MapperBuilder())->allowSuperfluousKeys()->mapper();
$result = $mapper->map(SomeClass::class, [
    'identifier' => [
        ['value' => 'some-identifier'],
        ['value' => 'some-identifier2']
    ],
    'description' => 'Lorem ipsum…',
]);
```

```
Fatal error: Uncaught CuyZ\Valinor\Mapper\TypeTreeMapperError: Could not map type `SomeClass`. An error occurred at path identifier.identifier: Value array{0: array{…}, 1: array{…}} is not a valid string
```

**Comment by hollow3464** [comment]
You can only create readonly properties through a constructor.

If you remove the readonly statement it will work. It will also work if you write a constructor that does the property assignments properly. 

When using a constructor, every property must me initialized or will remain uninitialized as the mapper will only use the constructor for variable assigment. Please refer to the language docs for usage on readonly props usage.

What is "weird" is needing to use the `@param` tag on the function phpDoc, else the mapper will recognize the array definition as `array<mixed>` and fail. Using `@var` above the parameter will not work. A working example would have `SomeClass` defined like this. 

```php
final class SomeClass
{
    public readonly string $description;

    /** @var array<Identifier> */
    public array $identifier;

    /**
     * @param array<Identifier> $identifier
     */
    public function __construct(
        string $description,
        array $identifier,
    ) {
        $this->description = $description;
        $this->identifier = $identifier;
    }
}
```

**Comment by romm** [comment]
Thanks for the report. @wengooooo this should be fixed with #497.

> What is "weird" is needing to use the @param tag on the function phpDoc, else the mapper will recognize the array definition as array<mixed> and fail. Using @var above the parameter will not work. A working example would have SomeClass defined like this.

@hollow3464 this behavior is intended, using `@var` works on properties, or when a constructor's parameter is also a promoted property:

```php
// NOT OK:
final class Foo
{
    public array $values;

    public function __construct(
        /** @var array<string> */
        array $values
    ) {
        $this->values = $values;
    }
}

// OK:
final class Foo
{
    public array $values;

    /**
     * @param array<string> $values
     */
    public function __construct(
        array $values
    ) {
        $this->values = $values;
    }
}

// OK:
final class Foo
{
    public function __construct(
        /** @var array<string> */
        public array $values
    ) {}
}
```
